docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,461 @@
+- Do we prefer `|json` or `| json` notation ?
+
+## LogQL Fields Operators
+
+Now that we have transformed our log streams into log fields we can pipe/chain a new set of operators.
+(Some operators are also compatible with log stream such as the `limit` and `uniq` operators.)
+
+For example the query below:
+
+```logql
+{job="prod/query-frontend"} | logfmt | filter level="info" and latency > 10 | sort duration desc | limit 25 | select query duration
+```
+
+returns the first 25 query and duration slower than 10s ordered by the slowest in the requested time range.
+
+Let's have a look in details how those new operators work.
+
+### Filter
+
+A filter operation noted `| filter` can filter out fields by their value using test predicates such as `path = "/foo/bar"`.
+
+Multiple test predicates can be chained/combined together using `and` and `or` logic operations.
+
+At least one leg of the test predicate must be a literal, this is to ensure we can infer the type to use for comparison. We could later add support for cross field comparison.
+
+__String literal__ will support the same Prometheus match type. (`=`,`=!`,`!~`,`=~`).
+
+Example: `{job="prod/query-frontend"} | logfmt | filter level=~"info|debug" and query_type!="filter"`
+
+__Numeric literal__ will support `=`,`!=` but also `<`,`<=`,`>`,`>=`.
+
+Example: `{job="prod/query-frontend"} | json | filter latency > 10 or status >= 400`
+
+When field value and literal types are different, conversion will be attempted prior testing and will be translated into zero type values in case of failure.
+
+In the future we will add more literals such as durations `2s` and dates (`03-03-2020T20:30`).
+
+### Sort
+
+The sort operation noted `| sort` allows to orders fields entries by a field value instead of the field entry time.
+
+> It should be noted that sorting is always executed after parsing fields entries in the time range. Not on the whole time range.
+
+For example: `{job="prod/query-frontend"} | json | sort latency desc`
+
+will sort  the first 1000 (default limit) fields entries by the highest latency first. Multiple fields can be used as fallback by the sort operation in case of equality.
+
+For example: `{job="prod/query-frontend"} | json | sort latency desc path asc` will sort by latency descending first then path ascending.
+
+Using sort operator with metric queries will result into an error.(e.g `count_over_time({job="prod/query-frontend"} | logfmt  | sort latency desc [5m])`)
+
+> __Warning:__ Sorting is dependent on how Grafana is currently ordering lines. Even if we re order our responses there is no guarantee that Grafana will be relying on it. And we might have to remove uncommon labels to keep only stream to ensure ordering across field streams. This means we will need a new API (`/v2`) to support ordering, although with compromise we might get something to work with the `v1` API.
```

Comment:
@davkal What do you think about this ?

Reply:
I think some feedback would be helpful here. How do other plugins (if any) implement upstream-defined sorting?

Reply:
Cloudwatch does support this I think, @kaydelaney ? This is a big document sorry ;)